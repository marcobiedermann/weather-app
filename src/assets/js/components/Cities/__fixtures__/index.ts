import cityFixtures from '../../City/__fixtures__';

const cities = [cityFixtures];

export default cities;
