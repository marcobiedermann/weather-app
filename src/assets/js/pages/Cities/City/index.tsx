import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import useSWR from 'swr';
import City from '../../../components/City';
import Loader from '../../../components/Loader';

interface Params {
  cityId: string;
}

interface CityPageQuery {
  coord: Coord;
  weather: Weather[];
  base: string;
  main: Main;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  sys: Sys;
  timezone: number;
  id: number;
  name: string;
  cod: number;
}

interface Clouds {
  all: number;
}

interface Coord {
  lon: number;
  lat: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  type: number;
  id: number;
  country: string;
  sunrise: number;
  sunset: number;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

const CityPage: FC = () => {
  const { cityId } = useParams<Params>();
  const { data, error } = useSWR<CityPageQuery>(`/weather?id=${cityId}`);

  if (error) {
    return <div>Error: {error}</div>;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>{data.name}</title>
      </Helmet>
      <City {...data} />
    </>
  );
};

export default CityPage;
