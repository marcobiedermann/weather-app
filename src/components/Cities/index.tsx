import React from 'react';
import CityPreview, { CityPreviewProps } from '../CityPreview';
import styles from './style.module.css';

interface CitiesProps {
  cities?: CityPreviewProps[];
}

function Cities(props: CitiesProps): JSX.Element {
  const { cities = [], ...otherProps } = props;

  return (
    <ul className={styles.cities} {...otherProps}>
      {cities.map((city) => (
        <li key={city.id}>
          <CityPreview {...city} />
        </li>
      ))}
    </ul>
  );
}

export type { CitiesProps };
export default Cities;
