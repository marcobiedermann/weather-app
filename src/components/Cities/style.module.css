.cities {
  display: grid;
  grid-gap: 1.5em;
  grid-template-columns: repeat(auto-fill, minmax(10em, 1fr));
  list-style: none;
  margin: 0;
  padding: 0;
}
