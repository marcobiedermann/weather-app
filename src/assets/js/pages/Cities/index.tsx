import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { Link, useRouteMatch } from 'react-router-dom';
import useSWR from 'swr';
import Cities from '../../components/Cities';
import Error from '../../components/Error';
import Loader from '../../components/Loader';
import { selectCities } from '../../reducers/cities';

interface CitiesPageQuery {
  cnt: number;
  list: List[];
}

interface List {
  coord: Coord;
  sys: Sys;
  weather: Weather[];
  main: Main;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  id: number;
  name: string;
}

interface Clouds {
  all: number;
}

interface Coord {
  lon: number;
  lat: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  country: string;
  timezone: number;
  sunrise: number;
  sunset: number;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

const CitiesPage: FC = () => {
  const cities = useSelector(selectCities);
  const { path } = useRouteMatch();
  const { t } = useTranslation();
  const { data, error } = useSWR<CitiesPageQuery>(`/group?id=${cities.join(',')}`);

  if (error) {
    return <Error message={error.message} />;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Cities</title>
      </Helmet>
      <Cities cities={data.list} />
      <Link to={`${path}/edit`}>{t('translation:edit')}</Link>
    </>
  );
};

export default CitiesPage;
