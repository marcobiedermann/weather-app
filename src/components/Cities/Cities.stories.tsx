import type { Meta, StoryObj } from '@storybook/react';
import Cities from '.';

const meta: Meta<typeof Cities> = {
  component: Cities,
  args: {
    cities: [
      {
        sys: {
          country: 'DE',
        },
        main: {
          temp: 18.54,
        },
        id: 2950159,
        name: 'Berlin',
      },
    ],
  },
};

type Story = StoryObj<typeof Cities>;

const Default: Story = {};

export { Default };
export default meta;
