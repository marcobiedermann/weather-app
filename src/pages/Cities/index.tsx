import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import Cities from '../../components/Cities';
import Error from '../../components/Error';
import Link from '../../components/Link';
import Loader from '../../components/Loader';
import Seo from '../../components/Seo';
import { useGroup } from '../../hooks';
import { selectCities } from '../../selectors/cities';
import { selectSettings } from '../../selectors/settings';
import { useAppSelector } from '../../store';

function CitiesPage(): JSX.Element {
  const { pathname } = useLocation();
  const navigate = useNavigate();
  const cities = useAppSelector(selectCities);
  const settings = useAppSelector(selectSettings);
  const { t } = useTranslation();
  const { data: group, error, isError, isPending } = useGroup(cities, settings);

  if (!cities.length) {
    navigate('/search');
  }

  if (isPending) {
    return <Loader />;
  }

  if (isError) {
    return <Error message={error.message} />;
  }

  return (
    <>
      <Seo title="Cities" />
      <Cities cities={group.list} />
      <Link to={`${pathname}/edit`}>{t('edit')}</Link>
    </>
  );
}

export default CitiesPage;
