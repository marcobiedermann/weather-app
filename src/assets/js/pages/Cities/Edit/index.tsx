import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useDispatch, useSelector } from 'react-redux';
import { Link, RouteComponentProps } from 'react-router-dom';
import useSWR from 'swr';
import Card from '../../../components/Card';
import City from '../../../components/CityPreview';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';
import { SEARCH } from '../../../constants/routes';
import { removeCity, selectCities } from '../../../reducers/cities';
import styles from './style.module.css';

interface EditPageQuery {
  cnt: number;
  list: List[];
}

interface List {
  coord: Coord;
  sys: Sys;
  weather: Weather[];
  main: Main;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  id: number;
  name: string;
}

interface Clouds {
  all: number;
}

interface Coord {
  lon: number;
  lat: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  country: string;
  timezone: number;
  sunrise: number;
  sunset: number;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

const EditPage: FC<RouteComponentProps> = () => {
  const dispatch = useDispatch();
  const cities = useSelector(selectCities);
  const { data, error } = useSWR<EditPageQuery>(`/group?id=${cities.join(',')}`);

  if (error) {
    return <Error message={error.message} />;
  }

  if (!data) {
    return <Loader />;
  }

  return (
    <>
      <Helmet>
        <title>Edit</title>
      </Helmet>
      <ul className={styles.cities}>
        {data.list.map((city) => (
          <li key={city.id}>
            <Card>
              <City {...city} />
              <button type="button" onClick={() => dispatch(removeCity(city.id))}>
                x
              </button>
            </Card>
          </li>
        ))}
        <li>
          <Card>
            <Link to={SEARCH.path}>Add</Link>
          </Card>
        </li>
      </ul>
    </>
  );
};

export default EditPage;
