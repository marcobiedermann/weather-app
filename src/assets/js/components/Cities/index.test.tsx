import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Cities from '.';
import citiesFixtures from './__fixtures__';

describe('components/Cities', () => {
  it('renders correctly', () => {
    expect.assertions(1);

    const tree = renderer
      .create(
        <MemoryRouter>
          <Cities cities={citiesFixtures} />
        </MemoryRouter>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
