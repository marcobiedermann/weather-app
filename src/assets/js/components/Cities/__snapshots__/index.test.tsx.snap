// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Cities renders correctly 1`] = `
<ul
  className="cities"
>
  <li>
    <div
      className="city"
      clouds={
        Object {
          "all": 0,
        }
      }
      coord={
        Object {
          "lat": 52.52,
          "lon": 13.41,
        }
      }
      dt={1605434959}
      visibility={10000}
      weather={
        Array [
          Object {
            "description": "clear sky",
            "icon": "01d",
            "id": 800,
            "main": "Clear",
          },
        ]
      }
      wind={
        Object {
          "deg": 170,
          "speed": 4.1,
        }
      }
    >
      <a
        className="city__link"
        href="/cities/2950159"
        onClick={[Function]}
      >
        <div
          className="city__temp"
        >
          12.61
           Â°
        </div>
        <h2
          className="city__name"
        >
          Berlin
        </h2>
        <h3
          className="city__country"
        >
          DE
        </h3>
      </a>
    </div>
  </li>
</ul>
`;
